### ISSUE TITLE
Something might be wrong with the materials tests?

### ISSUE THREAD

**Issue by keckler** [terrapower/armi#1144]
A [recent PR](https://github.com/terrapower/armi/pull/1114) was having some mysterious problems during testing:
- All tests were passing locally, even when running them in parallel. This same behavior was consistent for multiple users (myself and @john-science ).
- Tests triggered on GH were failing in a somewhat stochastic manner. Specifically, the version of python seemed to be making a difference in a stochastic way. Each time the test suite was run, some subset of the python versions would see failures. Sometimes it would be 3.7, other times 3.8, other times 3.10 (or some combination of multiple versions). But in each case that one of the versions would fail, the specific tests that failed were always the same. The root cause of these failures was never determined conclusively.

In an attempt to resolve this, a [new PR](https://github.com/terrapower/armi/pull/1140) was opened, which stepped through all the changes of the original PR systematically to nail down the problem area. In doing so,  the same kind of test failures started to be seen on [this commit](https://github.com/terrapower/armi/pull/1140/commits/bbae7421d434807dc61d1a08e16bcb2ecba4c719). That commit only added a single unit test to a new file. 

It was discovered that if the _same exact_ unit test was moved from that new file into the `test_materials.py` module, the stochastic failures would go away. The code implemented between the two PRs was ultimately exactly the same with this single exception.

Neither the reason for the initial stochastic failures nor the discovered fix were able to be understood. For this reason, it is suspected that a larger problem might exist somewhere deeper in the code, and this ticket is just here to document that possibility.

**Comment by john-science** [comment]
This error has popped up again in another PR that does nothing of interest to the bugs we see:  https://github.com/terrapower/armi/pull/1150

All that PR does is slightly change the density of copper.

So I assume the bug exists because of something in the unit tests.

I would say the bug could exist in `main`, but it only shows randomly. But so far the PRs that generate these errors have been VERY consistent at generating it. So it's not random.

**Comment by opotowsky** [comment]
>  Sometimes it would be 3.7, other times 3.8, other times 3.10 (or some combination of multiple versions). But in each case that one of the versions would fail, the specific tests that failed were always the same.

I think once one fails the rest get cancelled, so maybe this is just because of whichever one fails first

**Comment by opotowsky** [comment]
> This error has popped up again in another PR that does nothing of interest to the bugs we see: #1150
> 
> All that PR does is slightly change the density of copper.
> 
> So I assume the bug exists because of something in the unit tests.
> 
> I would say the bug could exist in `main`, but it only shows randomly. But so far the PRs that generate these errors have been VERY consistent at generating it. So it's not random.

This is interesting because the trigger is different. Chris' PR added a new unit test file, and when he moved the unit test to `test_materials.py` the error went away.

You just added a new test to `test_materials.py`, and are seeing the same error. The conclusion that there's some materials unit test gremlin makes sense, but it is unfortunate we can't figure out how to trigger it.

**Comment by keckler** [comment]
> > Sometimes it would be 3.7, other times 3.8, other times 3.10 (or some combination of multiple versions). But in each case that one of the versions would fail, the specific tests that failed were always the same.
> 
> I think once one fails the rest get cancelled, so maybe this is just because of whichever one fails first

What you're saying is true, but that's not the issue here. You can peek into the test logs to see which tests actually fail, versus the whole test run failing (or being cancelled). I've confirmed that way when tests actually fail or not.

**Comment by opotowsky** [comment]
Ah yes, I noticed the tests that were failing were the same ones. 

So you're saying, e.g., the python 3.7 run would pass and the python 3.8 run would fail?

**Comment by john-science** [comment]
The extra annoying bit is I can't reproduce this nonsense locally.

So the turn-around time on any test is huge.

**Comment by john-science** [comment]
So... I did a manual `git bisect`, and ran the "minor change to copper, plus a single unit test" change from [this PR](https://github.com/terrapower/armi/pull/1150) through a ton of old commits.  The first commit in ARMI `main` history that exhibited this error was:

[PR #1043](https://github.com/terrapower/armi/commit/12739560b6947a4a95f9bd199cde614b03a9a98f#diff-cc777c5960e4ddaebe2331d8bfd69890beac9f578def9b04f1738f3ea4d1f8dd)

So, if this error is deterministic, it was caused by that PR. Or maybe that PR with #1022, which occured before it.

This would be great news, especially since that PR touches materials and `test_material.py`.  But I still don't see it. Mostly, that PR just removes unused code and does a little refactoring.

**Comment by john-science** [comment]
So, after extensive A/B testing and git bisecting, I determined that removing these two tests from the `_Material_Test` baseclass solves the problem:

https://github.com/terrapower/armi/blob/5cfddf607691184eb407c5ea8a55c9f31396d8fd/armi/materials/tests/test_materials.py#L52-L56

Unfortunately, I have no solid explanation as to why. yet.

**Comment by john-science** [comment]
Okay, so the change that caused this testing issue was when I [removed the `Leaf` class](https://github.com/terrapower/armi/pull/1043/files#diff-7df757771965f6a52296b110d7d4b9224007868d636955c89ce53daa55d45396) from ARMI.

One side-effect of this nothing change is I set the `children` of the `Material` class equal to empty `[]`, because it no longer made sense for _materials_ in ARMI to have geometry objects which were their children.  Which is fine.

The problem came when I tried to verify this in the material test base class `_Material_Test` for all materials.  Apparently, when run in parallel, we still have a couple of unit tests which will drop data because of this. I can't reproduce it locally, with any version of Python, but it happens very reliably on GitHub.

I still am not 100%, but the problem usually seems to occur around this piece of code:

https://github.com/terrapower/armi/blob/2a9cd5c0c0f656259d1e3dd48f160c58e3ff5961/armi/reactor/composites.py#L1998-L2010

**Comment by john-science** [comment]
This problem is solved. It's to do with a really github-specific race condition, when we note all the children of a material are empty.  

This was really persnickety to find, and though I can explain it, it is a race condition that I just _can't_ reproduce locally.

**Comment by keckler** [comment]
Was anything changed to resolve this? I didn't see any changes made, so I'm confused why its being closed.

Is this problem still going to possibly happen when new PRs are put up?
